src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -236,27 +243,103 @@ public bool RefreshOption(OptionKey2 optionKey, object? newValue)
             return true;
         }
 
+        public void AddOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+        {
+            lock (_gate)
+            {
+                if (_keepAliveTable.TryGetValue(target, out var existingHandler))
+                {
+                    var newHandler = existingHandler + handler;
+#if NET6_0_OR_GREATER
+                    _keepAliveTable.AddOrUpdate(target, newHandler);
+#else
+                    _keepAliveTable.Remove(target);
+                    _keepAliveTable.Add(target, newHandler);
+#endif
+                }
+            }
+
+            ImmutableInterlocked.Update(
+                ref _weakHandlers,
+                static (weakHandlers, handler) =>
+                {
+                    return weakHandlers
+                        .RemoveAll(WeakReferenceExtensions.IsNull)
+                        .Add(new WeakReference<EventHandler<OptionChangedEventArgs>>(handler));
+                },
+                handler);
+        }
+
+        public void RemoveOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
```

Comment:
Can we abstract this logic to something like `WeakEvent<T>`?

Reply:
➡️ Extracted a helper type `WeakEvent<TEventArgs>`. It's not _always_ the right thing to use, but for global events where we control the API, the `AddHandler` and `RemoveHandler` methods are much easier to use than types that try to extract the target from the event handler delegate itself.